substrate/frame/nomination-pools/src/migration.rs

```diff
@@ -107,6 +107,115 @@ pub mod unversioned {
 			Ok(())
 		}
 	}
+
+	/// Migrate existing pools from [`adapter::StakeStrategyType::Transfer`] to
+	/// [`adapter::StakeStrategyType::Delegate`].
+	///
+	/// Note: This only migrates the pools, the members are not migrated. They can use the
+	/// permission-less [`Pallet::migrate_delegation()`] to migrate their funds.
+	///
+	/// This migration does not break any existing pool storage item, does not need to happen in any
+	/// sequence and hence can be applied unversioned on a production runtime.
+	pub struct DelegationStakeMigration<T>(sp_std::marker::PhantomData<T>);
+
+	impl<T: Config> OnRuntimeUpgrade for DelegationStakeMigration<T> {
+		fn on_runtime_upgrade() -> Weight {
+			if StrategyMigration::<T>::get() == Some(adapter::StakeStrategyType::Delegate) {
+				log!(info, "`pools::migration::unversioned::DelegationStakeMigration` already migrated and can be removed.");
+				return T::DbWeight::get().reads_writes(1, 0)
+			}
+
+			let mut migrate_count: u32 = 0;
+			let mut fail_count: u32 = 0;
+			BondedPools::<T>::iter().for_each(|(id, _inner)| {
```

Comment:
This can still be attacked by creating many pools right before the migration... 

I don't think there is an easy and sound way to fix it. What comes to mind is `.take(DEFENSIVE_MAX_POOLS_I_KNOW)`. So we don't iterate too much and the chain won't brick, but then you are stuck in a situation when you have some pools migrated and som unmigrated.. 

This is only safe if there was a `MaxPools`, but we don't have one set in Polkadot. 

It seems to be that a one-off MBM that does pools + members is a bit more work in advance, but less work on agregate.

Reply:
Unless, if: 

1. it is okay to have some pools migrated and some not
2. add a permissionelss transaction to also migrate any leftover pools 
3. run this migration with a `.take(SAFE_BOUND)`

So this migration is just a best effort, and even pools should be migrated lazily.

Reply:
Yeah good point. Was discussing with Goncalo as well about having a permissionless extrisinsic for the unlikely scenario where some pools fail to migrate (should really not happen though). But your suggestion does play well with it.

Practically though, we need 500 DOT to create a pool so it would be a very expensive attack which will only cause the migration block to be longer. :D